src/Text/Parsing/Parser/String.purs

```diff
@@ -2,89 +2,113 @@
 
 module Text.Parsing.Parser.String where
 
-import Data.String as S
-import Control.Monad.State (modify, gets)
 import Data.Array (many)
-import Data.Foldable (elem, notElem)
+import Data.Foldable (fold, elem, notElem)
+import Data.List as L
+import Data.Monoid.Endo (Endo(..))
 import Data.Maybe (Maybe(..))
-import Data.Newtype (wrap)
-import Data.String (Pattern, fromCharArray, length, singleton)
+import Data.Newtype (class Newtype, unwrap)
+import Data.String as S
+import Control.Monad.State (modify, gets)
 import Text.Parsing.Parser (ParseState(..), ParserT, fail)
 import Text.Parsing.Parser.Combinators (try, (<?>))
-import Text.Parsing.Parser.Pos (updatePosString)
+import Text.Parsing.Parser.Pos (Position, updatePosString, updatePosChar)
 import Prelude hiding (between)
 
+-- | A newtype used in cases where there is a prefix to be stipPrefixed.
+newtype Prefix a = Prefix a
+
+derive instance eqPrefix :: Eq a => Eq (Prefix a)
+derive instance ordPrefix :: Ord a => Ord (Prefix a)
+derive instance newtypePrefix :: Newtype (Prefix a) _
+
+instance showPrefix :: Show a => Show (Prefix a) where
+  show (Prefix s) = "(Prefix " <> show s <> ")"
+
+class HasUpdatePosition a where
+  updatePos :: Position -> a -> Position
+
+instance stringHasUpdatePosition :: HasUpdatePosition String where
+  updatePos = updatePosString
+
+instance charHasUpdatePosition :: HasUpdatePosition Char where
+  updatePos = updatePosChar
+
 -- | This class exists to abstract over streams which support the string-like
 -- | operations which this modules needs.
-class StringLike s where
-  drop :: Int -> s -> s
-  indexOf :: Pattern -> s -> Maybe Int
-  null :: s -> Boolean
-  uncons :: s -> Maybe { head :: Char, tail :: s }
-
-instance stringLikeString :: StringLike String where
-  uncons = S.uncons
-  drop = S.drop
-  indexOf = S.indexOf
-  null = S.null
-
--- | Match end-of-file.
-eof :: forall s m. StringLike s => Monad m => ParserT s m Unit
+-- |
+-- | Instances must satisfy the following laws:
+-- | - `stipPrefix (Prefix a) a >>= uncons = Nothing`
+class StreamLike f c | f -> c where
+  uncons :: f -> Maybe { head :: c, tail :: f, updatePos :: Position -> Position }
+  stipPrefix :: Prefix f -> f -> Maybe {  rest :: f, updatePos :: Position -> Position }
+
+instance stringStreamLike :: StreamLike String Char where
+  uncons f = S.uncons f <#> \({ head, tail}) ->
+    { head, tail, updatePos: (_ `updatePos` head)}
+  stipPrefix (Prefix p) s = S.stripPrefix (S.Pattern p) s <#> \rest ->
+    { rest, updatePos: (_ `updatePos` p)}
+
+instance listcharStreamLike :: (Eq a, HasUpdatePosition a) => StreamLike (L.List a) a where
```

Comment:
will do once this is merged https://github.com/purescript/purescript-lists/pull/126